contracts/IndexStaking.sol

```diff
@@ -0,0 +1,61 @@
+//SPDX-License-Identifier: Unlicense
+pragma solidity ^0.6.6;
+
+import "@openzeppelin/contracts/math/SafeMath.sol";
+import "../libraries/SafeTransfer.sol";
+
+contract IndexStaking {
+    using SafeMath for uint256;
+
+    uint public constant duration = 3 * 365 * 6500; // blocks
+    uint public constant totalSupply = 100000000 * 10**18;
+    uint public constant rewardSupply = 48 * totalSupply / 100;
+
+    address public stakingToken;
+    address public rewardToken;
+
+    uint public staked; // S
+    uint public activeStakeDeposits; // T
+    mapping(address => uint) public stake; // stake
+    mapping(address => uint) public stakedSnapshot; // S0
+
+    uint public accumulatedReward;
+    uint public startBlock;
+
+    constructor(address _stakingTokenAddress, address _rewardTokenAddress) public {
+        stakingToken = _stakingTokenAddress;
+        rewardToken = _rewardTokenAddress;
+        startBlock = block.number;
+    }
+
+    function deposit(uint _amount) external {
+        SafeTransfer.transferFromERC20(address(stakingToken), msg.sender, address(this), _amount);
+        stake[msg.sender] = _amount;
```

Comment:
This would overwrite user's existing stake. Make sure to prevent this from happening.